pkg/roachprod/vm/gce/gcloud.go

```diff
@@ -473,6 +480,164 @@ func (p *Provider) GetVMSpecs(
+	syncutil.Mutex
+	cache map[string]*MachineTypeSpecs
+}
+
+type MachineTypeSpecs struct {
+	Architecture                 string
+	Description                  string
+	GuestCpus                    int
+	Kind                         string
+	MaximumPersistentDisks       int
+	MaximumPersistentDisksSizeGb string
+	MemoryMb                     int
+	Name                         string
+	Zone                         string
+}
+
+type CacheKeyLookup string
+
+const (
+	NarrowCacheKey CacheKeyLookup = "NARROW_CACHE_KEY"
+	WideCacheKey   CacheKeyLookup = "WIDE_CACHE_KEY"
+)
+
+func (mts *ProviderMachineTypeSpecsCache) get(
+	machineType, project, zone string, ckl CacheKeyLookup,
+) (*MachineTypeSpecs, error) {
+
+	mts.Lock()
+	defer mts.Unlock()
+
+	if mts.cache == nil {
+		mts.cache = make(map[string]*MachineTypeSpecs)
+	}
+
+	// We require an "{instance-type}-{project}-{zone} combination to describe
+	// the instance type (as required by GCP), but store the cached results
+	// under two different keys:
+	// - a narrow one "{instance-type}-{project}-{zone}"
+	// - a wide one "{instance-type}-{project}"
+	// The justification behind this is that instance types are homogeneous
+	// across all zones, so we can increase cache hits when we don't actually
+	// care about the zone, but we can also request a narrow cache key when
+	// we want to ensure that this instance type is available in a specific zone
+	buildCacheKeys := func(machineType, project, zone string) map[CacheKeyLookup]string {
+		return map[CacheKeyLookup]string{
+			WideCacheKey:   machineType + "-" + project,
+			NarrowCacheKey: machineType + "-" + project + "-" + zone,
+		}
+	}
+	cacheKeys := buildCacheKeys(machineType, project, zone)
+
+	if cachedSpecs, ok := mts.cache[cacheKeys[ckl]]; ok {
+		return cachedSpecs, nil
+	}
+
+	// Query GCP to get machine type specs:
+	// Running:
+	// 		gcloud compute machine-types describe t2a-standard-4 \
+	//  		--project cockroach-ephemeral --zone us-east1-b \
+	// 			--format json
+	//
+	// Note: GCP requires a zone to describe a machine type, though during
+	// cluster creation, roachprod's takes a `--gce-machine-type` as input and
+	// guesses the CPU architecture from it. Only then, the zone is inferred
+	// from an hardcoded CPU architecture/zone mapping.
+	// As we're want to know the architecture from the machine type before
+	// knowing the zone, we try the default zone for both amd64 and arm64
+	// architectures.
+	//
+	// In cases where we know the zone already (e.g. `roachprod list`), we use
+	// the supplied zone to query, making sure that we won't hit a zone in which
+	// the instance type is not available.
+
+	var zonesIteration []string
+	if zone != "" {
+		zonesIteration = []string{zone}
+	} else {
+		zonesIteration = []string{
+			DefaultZones(string(vm.ArchAMD64))[0],
+			DefaultZones(string(vm.ArchARM64))[0],
+		}
+	}
+
+	found := false
+	specs := &MachineTypeSpecs{}
+	for _, zone := range zonesIteration {
```

Comment:
We can parallelise this call to ensure `roachprod/roachtest` remains speedy.

Reply:
This should actually never happen as `roachprod` syncs before creating, so with the use of the `WideCacheKey` probability that we already fetched the specs for this machine-type in a zone is high. That being said, I implemented parallelization because you're right... why risk a slow down when it's easy not to!

On the note of speediness, querying the machine types does slow down `sync` quite a bit. From my initial tests, it makes a `roachprod list` go from ~10s to ~15s. Without the caching mechanism, that goes to close to 30s.